### PR TITLE
Add openapi3 spec

### PR DESCRIPTION
**Pull request by Drogglbecher**
closes https://github.com/davishmcclurg/json_schemer/issues/22

### CODE REVIEW

**Comment by ahx** [comment]
@davishmcclurg Do you think we should base this on Draft4 or Draft7? – https://github.com/OAI/OpenAPI-Specification/blob/master/versions/3.0.2.md#schemaObject is not clear about that.

Personally I would like to change this to extend Draft7, because it has useful features like "date" format.

**Comment by davishmcclurg** [comment]
Sorry for the delay on reviewing this. I was on vacation for last couple weeks. I only had time for a quick look right now, but it looks great! I'll review more closely when I get some time this week. I'll also do some reading on the openapi spec to determine what draft to use, @ahx.

**Comment by ahx** [comment]
No problem. Thanks! We did this during an internal hackathon so there should be some lines to polish in there. 
As far as specification goes, ‘nullable’ and ‘discriminator’ seem to be the only new features of the OAS flavor, but we might have missed something there. So take your time and don’t hesitate to leave comments with questions here so we can help trying to understand the spec.

File: lib/json_schemer/schema/base.rb [hunk 1]
```diff
@@ -54,10 +54,29 @@ def validate(data)
 
     protected
 
+      def discriminate(instance, field_name)
```
Comment (dleavitt) [inline lib/json_schemer/schema/base.rb]: wrong number of args here. should add a test like this to check that using OA attributes in other schemas doesn't cause them to break, e.g:
```ruby
  def test_do_unknown_properties_break_other_schemas
    schema = JSONSchemer.schema(
      'discriminator' => 'yup', # or any truthy value
      "oneOf" => [{'type' => 'object'}],
    )
    assert schema.valid?('a' => 1)
  end
```

File: test/json_schemer_test.rb [hunk 2]
```diff
@@ -529,23 +529,54 @@ def test_it_returns_nested_errors
...
-    Dir["JSON-Schema-Test-Suite/tests/#{version}/**/*.json"].each_with_index do |file, file_index|
+
+    paths = []
+
+    if version == 'openapi3'
+      version = 'draft4'
+      paths += Dir['test/openapi3/**/*.json']
+    end
+
+    paths += Dir['JSON-Schema-Test-Suite/tests/#{version}/**/*.json']
```
[the remark this answers is not in the document]
Comment (ahx) [reply test/json_schemer_test.rb]: @dleavitt Thanks. Should be fixed via https://github.com/davishmcclurg/json_schemer/pull/30/commits/721cd3d6ed2d259ff9ee3bb6786e4737331a564d

File: lib/json_schemer/schema/base.rb (hunk 1 above)
Comment (ahx) [reply lib/json_schemer/schema/base.rb]: @dleavitt Thanks again. I have moved the method out of Base, because it  does not make sense there. 👍

**Comment by davishmcclurg** [comment]
From what I can tell, OpenAPI 3 is based on JSON Schema Draft 5 (draft-wright-json-schema-00), which is basically draft 4. Unfortunately, there's a long list of deviations and additions: https://github.com/OAI/OpenAPI-Specification/blob/master/versions/3.0.2.md#schema-object

I wasn't able to find a test suite similar to the JSON Schema one (https://github.com/json-schema-org/JSON-Schema-Test-Suite), which makes me uncomfortable merging this because there are probably inconsistencies. I'd rather not take on supporting another specification if I'm not reasonably confident we're compliant.

Thanks for your work on the PR. I'm going to leave it open for discussion, but at this point I'm not planning to merge.

**Pull request closed without merging**